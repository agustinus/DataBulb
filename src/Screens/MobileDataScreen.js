import React from 'react';
import {View, SafeAreaView, StyleSheet} from 'react-native';
import {connect} from 'react-redux';
import {ActionCreator} from '../Redux/DataGovAction';

import LoadingIndicator from '../Components/LoadingIndicator';
import MobileDataItemView from '../Components/MobileDataItemView';
import StandardList from '../Components/StandardList';
import VerticalTitle from '../Components/VerticalTitle';
import GeneralModal from '../Components/GeneralModal';

export class MobileDataScreen extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      flatListReady: false,
      showModal: false,
      modalContent: null,
    };
  }

  componentDidMount() {
    this.props.fetchMobileDataUsage();
  }

  render() {
    return (
      <SafeAreaView style={styles.safeArea}>
        <View style={styles.screen}>
          <View style={styles.title}>
            <VerticalTitle />
          </View>
          <View style={styles.listWrapper}>
            {this.props.data ? (
              <StandardList
                style={styles.standardList}
                data={this.props.data}
                renderItem={({item}) => this._renderItem({item})}
                isRefreshing={this.props.isLoading}
                onEndReached={() => this._loadMore()}
                onScroll={this._scrolled.bind(this)}
                onRefresh={() => this.props.refreshMobileDataUsage()}
              />
            ) : (
              undefined
            )}
          </View>
        </View>
        {this.props.error ? this._showError(this.props.error) : undefined}
        {this.props.isLoading ? (
          <LoadingIndicator isLoading={this.props.isLoading} />
        ) : (
          undefined
        )}
        {this.state.modalContent ? (
          <GeneralModal
            isVisible={true}
            onDismiss={() => this._onDismissModal()}
            title={'Usage decreased in '}
            content={this._buildModalContent(
              this.state.modalContent.decreasedQuarter,
            )}
          />
        ) : (
          undefined
        )}
      </SafeAreaView>
    );
  }

  _showError(error) {
    const isNetworkError =
      !error.status &&
      error.message &&
      error.message === 'Network request failed';
    return (
      <GeneralModal
        isVisible={true}
        onDismiss={() => {
          isNetworkError ? this.props.fetchMobileDataUsage() : null;
        }}
        title={isNetworkError ? error.message : "Oops, It's error"}
        content={error.status ? this.props.error.status.toString() : 'ðŸ˜¢'}
        buttonText={isNetworkError ? 'REFRESH' : ''}
      />
    );
  }

  _renderItem = ({item}) => {
    return (
      <MobileDataItemView
        year={item.year}
        total={item.total}
        showIcon={item.decreasedQuarter.length > 0}
        onIconClicked={() => {
          this.setState({modalContent: item});
        }}
      />
    );
  };

  _onDismissModal() {
    this.setState({modalContent: null});
  }

  _buildModalContent(decreasedQuarters) {
    let content = '';
    decreasedQuarters.forEach(quarter => {
      if (content) {
        content += ' ';
      }
      content += 'Q' + quarter;
    });
    return content;
  }

  _scrolled = () => {
    if (this.state.flatListReady) {
      return null;
    }
    this.setState({flatListReady: true});
  };

  _loadMore = () => {
    if (this.props.isCompleted) {
      return null;
    }
    this.props.fetchMobileDataUsage();
  };
}

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: '#fff',
  },
  screen: {
    flex: 1,
    alignItems: 'stretch',
    justifyContent: 'center',
    flexDirection: 'row',
  },
  title: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  listWrapper: {
    flex: 8,
  },
  standardList: {
    flex: 8,
    paddingLeft: 8,
    paddingRight: 15,
    paddingTop: 5,
  },
  listItemTitle: {
    fontWeight: 'bold',
  },
});

const mapStateToProps = (state /*, ownProps*/) => {
  return {
    isLoading: state.dataGov.isFetching,
    data: state.dataGov.data,
    error: state.dataGov.error,
    offset: state.dataGov.offset,
    isCompleted: state.dataGov.isCompleted,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    fetchMobileDataUsage() {
      return dispatch(ActionCreator.fetchMobileDataUsage());
    },
    refreshMobileDataUsage() {
      return dispatch(ActionCreator.refreshMobileDataUsage());
    },
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(MobileDataScreen);
