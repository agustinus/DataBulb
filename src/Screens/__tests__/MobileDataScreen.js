import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import {configure, shallow} from 'enzyme';
import {MobileDataScreen} from '../MobileDataScreen';

configure({adapter: new Adapter()});

describe('MobileDataScreen', () => {
  let wrapper = null;
  let instance = null;

  beforeEach(() => {
    const props = {
      fetchMobileDataUsage: jest.fn(),
      data: [{year: '2000', total: '0.000123', decreasedQuarter: []}],
    };
    wrapper = shallow(<MobileDataScreen {...props} />);
    instance = wrapper.instance();
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('StandardList').length).toBe(1);
  });

  it('should render item correctly', () => {
    const item = instance._renderItem({
      item: {year: '2000', total: '0.000123', decreasedQuarter: []},
    });
    expect(item).toMatchSnapshot();
  });

  it('should invoke _loadMore() if onEndReached triggered', () => {
    wrapper.setState({flatListReady: true, isCompleted: false});
    wrapper
      .find('StandardList')
      .props()
      .onEndReached();
    expect(instance.props.fetchMobileDataUsage).toHaveBeenCalledTimes(2);
  });

  it('should invoke _renderItem() if renderItem triggered', () => {
    const item = wrapper
      .find('StandardList')
      .props()
      .renderItem({
        item: {year: '2000', total: '0.000123', decreasedQuarter: []},
      });
    expect(item).toMatchSnapshot();
  });

  it('should set modalContent state with the item when onIconClicked is triggered', () => {
    const item = instance._renderItem({
      item: {year: '2000', total: '0.000123', decreasedQuarter: [1]},
    });
    const component = shallow(item);
    component
      .find('TouchableOpacity')
      .props()
      .onPress();
    expect(wrapper.state().modalContent).toEqual({
      year: '2000',
      total: '0.000123',
      decreasedQuarter: [1],
    });
  });

  it('should return modal content', () => {
    const content = instance._buildModalContent([1, 2, 3]);
    expect(content).toEqual('Q1 Q2 Q3');
  });

  it('should clear state when _onDismissModal is called', () => {
    const state = {decreasedQuarter: [1, 2]};
    wrapper.setState({modalContent: state});
    expect(wrapper.state().modalContent).toEqual(state);
    instance._onDismissModal();
    expect(wrapper.state().modalContent).toBeNull();
  });

  it('should show modal', () => {
    expect(wrapper.find('GeneralModal').length).toBe(0);
    wrapper.setState({modalContent: {decreasedQuarter: [1, 2]}});
    // console.log('TEST: ', comp.debug({verbose: true}));
    expect(wrapper.find('GeneralModal').length).toBe(1);
  });

  it('should invoke _onDismissModal when modal is dismissed', () => {
    expect(wrapper.find('GeneralModal').length).toBe(0);
    wrapper.setState({modalContent: {decreasedQuarter: [1, 2]}});
    wrapper
      .find('GeneralModal')
      .props()
      .onDismiss();
    expect(wrapper.state().modalContent).toBeNull();
  });

  it('should set state flatListReady only one when it is scrolled', () => {
    wrapper.setState({flatListReady: false});
    expect(wrapper.state().flatListReady).toEqual(false);
    instance._scrolled();
    expect(wrapper.state().flatListReady).toEqual(true);
    const result = instance._scrolled();
    expect(result).toBeNull();
    expect(wrapper.state().flatListReady).toEqual(true);
  });

  it('should call fetchMobileDataUsage (_loadMore) if flatlist is ready and still have next data', () => {
    wrapper.setProps({isCompleted: true});
    instance._loadMore();
    expect(instance.props.fetchMobileDataUsage).toHaveBeenCalledTimes(1);

    wrapper.setProps({isCompleted: false});
    instance._loadMore();
    expect(instance.props.fetchMobileDataUsage).toHaveBeenCalledTimes(2);
  });

  it('should show error message for Network request failed', () => {
    wrapper.setProps({error: {message: 'Network request failed'}});
    const item = wrapper.find('GeneralModal');
    expect(item).toMatchSnapshot();
  });

  it('should handle refresh button for Network request failed', () => {
    wrapper.setProps({error: {message: 'Network request failed'}});
    const item = wrapper.find('GeneralModal');
    item.props().onDismiss();
    expect(instance.props.fetchMobileDataUsage).toHaveBeenCalledTimes(2);
  });

  it('should show error message for http error', () => {
    wrapper.setProps({error: {status: 500}});
    const item = wrapper.find('GeneralModal');
    expect(item).toMatchSnapshot();
  });

  it('should do nothing when OK button is pressed for http error', () => {
    wrapper.setProps({error: {status: 500}});
    const item = wrapper.find('GeneralModal');
    console.log('TEST: ', item.debug({verbose: true}));
    expect(item.props().onDismiss()).toBeUndefined();
  });

  it('should show loading indicator', () => {
    expect(wrapper.find('LoadingIndicator').exists()).toBe(false);
    wrapper.setProps({isLoading: true});
    expect(wrapper.find('LoadingIndicator').exists()).toBe(true);
  });
});
