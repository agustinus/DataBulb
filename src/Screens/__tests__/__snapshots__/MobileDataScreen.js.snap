// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileDataScreen renders correctly 1`] = `ShallowWrapper {}`;

exports[`MobileDataScreen should invoke _renderItem() if renderItem triggered 1`] = `
<MobileDataItemView
  onIconClicked={[Function]}
  showIcon={false}
  total="0.000123"
  year="2000"
/>
`;

exports[`MobileDataScreen should render item correctly 1`] = `
<MobileDataItemView
  onIconClicked={[Function]}
  showIcon={false}
  total="0.000123"
  year="2000"
/>
`;
